### PR TITLE
Mag cdf

### PR DESCRIPTION
**Pull request by maxinelasp**
# Change Summary
Updates to MAG to add raw output and initial pass at CDF files.

## Overview

Added CDF files plus a few methods to create them

## New Files

- mag_l1a
   - Call methods for decom
- mag_cdf_attrs
   -  shared instrument attributes for CDF files
   
## Updated Files

- global cdf attrs
   - Created system to add an instrument PI

### CODE REVIEW

File: imap_processing/cdf/global_attrs.py
Comment (greglucas) [inline imap_processing/cdf/global_attrs.py]: It looks like PI_NAME should only be one entry according to the CDF docs. Should this be replaced with the instrument PI instead of appended to with both the mission PI and instrument PI?

@bryan-harter would be good for you to review these CDF changes to see if there is a standard way of putting this into the files that SPDF approves of.

File: imap_processing/cdf/global_attrs.py
Comment (greglucas) [inline imap_processing/cdf/global_attrs.py]: There should be a space before the colon in numpydoc for all the parameters in the docstrings https://numpydoc.readthedocs.io/en/latest/format.html#parameters
```suggestion
    instrument_type : str, default="Particles (space)"
```

File: imap_processing/mag/l0/decom_mag.py
Comment (greglucas) [inline imap_processing/mag/l0/decom_mag.py]: Why don't we want warning messages? I'd just leave this alone, but we should only update this logger and not the global level.
```suggestion
logger = logging.getLogger(__name__)
logger.setLevel(logging.ERROR)
```

File: imap_processing/mag/l0/decom_mag.py
Comment (greglucas) [inline imap_processing/mag/l0/decom_mag.py]: Speaking with Gavin earlier in that thread on slack we shouldn't be updating this, and in fact it makes it much slower due to the buffer thrashing issues on space_packet_parser. We can safely use the default.
```suggestion
            mag_packets = mag_parser.generator(binary_data)
```

File: imap_processing/mag/l0/decom_mag.py
Comment (greglucas) [inline imap_processing/mag/l0/decom_mag.py]: Can you remove this try/except altogether? I believe I provided some code that should fix this. IMO we have a bad packet definition if we are running into this. I'm happy to open a separate PR with that code, just let me know if you want me to do that separately.

File: imap_processing/mag/l0/decom_mag.py
Comment (greglucas) [inline imap_processing/mag/l0/decom_mag.py]: Is this a placeholder for now to be filled in with real data later? Or is this a coordinate that is "fixed" in a sense? I was expecting to see something related to `norm_data["raw_vectors"]`, so it might help to put a quick comment in one way or the other.

File: imap_processing/mag/l0/decom_mag.py
Comment (greglucas) [inline imap_processing/mag/l0/decom_mag.py]:
```suggestion
```

File: imap_processing/mag/l0/decom_mag.py
Comment (greglucas) [inline imap_processing/mag/l0/decom_mag.py]: Do you want this print there? If so, I'd add some extra context, otherwise if it is debugging I'd remove it and leave the TODO.

File: imap_processing/mag/l1a/mag_l1a.py
Comment (greglucas) [inline imap_processing/mag/l1a/mag_l1a.py]: There is no "mode" anymore, so you might need a rebase to get the updates. I think this would be "descriptor" now.

https://github.com/IMAP-Science-Operations-Center/imap_processing/blob/497d446ebfe7a6eea45dc6d85e397338708741e4/imap_processing/cdf/utils.py#L41-L44

File: imap_processing/mag/mag_cdf_attrs.py
Comment (greglucas) [inline imap_processing/mag/mag_cdf_attrs.py:18]: That page is frustrating to read with the jumping header :) Good job having the patience to grab that text!

File: imap_processing/mag/tests/mag_multiple_packets.pkts
Comment (greglucas) [inline imap_processing/mag/tests/mag_multiple_packets.pkts:1]: Curious in GitHub's UI these show as new files. Are they newly added here, or just updated?
These files are in the base repo already it looks like: https://github.com/IMAP-Science-Operations-Center/imap_processing/tree/dev/imap_processing/tests/mag

File: imap_processing/cdf/global_attrs.py
Comment (maxinelasp) [reply imap_processing/cdf/global_attrs.py]: Yeah, I wasn't sure if it should include the instrument PI or not. The example files I had from MAG did.

File: imap_processing/mag/l0/decom_mag.py
Comment (maxinelasp) [reply imap_processing/mag/l0/decom_mag.py]: Oh yeah, this is left over from something else. Let me take it out.

File: imap_processing/mag/l0/decom_mag.py
Comment (maxinelasp) [reply imap_processing/mag/l0/decom_mag.py]: Yup, just missed updating it here, I will remove!

File: imap_processing/mag/l0/decom_mag.py
Comment (maxinelasp) [reply imap_processing/mag/l0/decom_mag.py]: Ah yeah that was left in for future work, but shouldn't be merged. I'll remove it and add it in the TODO or my notes

File: imap_processing/mag/l1a/mag_l1a.py
Comment (maxinelasp) [reply imap_processing/mag/l1a/mag_l1a.py]: Yeah, this was a merge fail! Or de-merge? I deleted the commit that did this. I'll update it to match the current version of write_cdf in the code

File: imap_processing/mag/mag_cdf_attrs.py
Comment (maxinelasp) [reply imap_processing/mag/mag_cdf_attrs.py:18]: Lol I hate the website design of that page, it's so jerky

File: imap_processing/mag/tests/mag_multiple_packets.pkts
Comment (maxinelasp) [reply imap_processing/mag/tests/mag_multiple_packets.pkts:1]: Ah, I know what the problem is. These are under /mag/tests, when they should be (and are in dev) under tests/mag. I'll make sure the test file under tests/mag has everything and then remove these.

File: imap_processing/mag/l1a/mag_l1a.py
Comment (maxinelasp) [reply imap_processing/mag/l1a/mag_l1a.py]: Also, this method signature is changed again in my other CLI PR, just so you're aware.

File: imap_processing/mag/tests/mag_multiple_packets.pkts
Comment (maxinelasp) [reply imap_processing/mag/tests/mag_multiple_packets.pkts:1]: Also I apparently failed to write tests for CDF stuff. I'll write a few mag-specific ones as well before merging.

Comment (greglucas) [reply imap_processing/mag/tests/mag_multiple_packets.pkts:1]: Interesting! I think there is something confusing GitHub somewhere because now it shows them being deleted even though they aren't on the main branch :)

I _think_ once you squash down to a single commit on merge this will resolve itself...

File: imap_processing/mag/l0/decom_mag.py
Comment (greglucas) [reply imap_processing/mag/l0/decom_mag.py]: Awesome, thanks!

File: imap_processing/mag/l1a/mag_l1a.py
Comment (greglucas) [reply imap_processing/mag/l1a/mag_l1a.py]: :+1:

File: imap_processing/mag/tests/mag_multiple_packets.pkts
Comment (greglucas) [reply imap_processing/mag/tests/mag_multiple_packets.pkts:1]: Ignore me! I was only reviewing the single commit. Would be nice to squash down still though to avoid added/removed files. Thanks for figuring that out!

File: imap_processing/cdf/global_attrs.py
Comment (sdhoyt) [inline imap_processing/cdf/global_attrs.py]: Is the purpose of appending to PI names/affiliations defined in this file to add the instrument PI?

File: imap_processing/cdf/global_attrs.py
Comment (sdhoyt) [inline imap_processing/cdf/global_attrs.py:8]: @bryan-harter This isn't something we'd want to change in this PR, but are there any attributes we should add in here to conform to SPASE?

File: imap_processing/mag/l0/decom_mag.py
Comment (sdhoyt) [inline imap_processing/mag/l0/decom_mag.py:46]: Could you use the decom.decom_packets function here?

File: imap_processing/mag/tests/mag_multiple_packets.pkts
Comment (maxinelasp) [reply imap_processing/mag/tests/mag_multiple_packets.pkts:1]: Yeah definitely will squash, don't worry 😆 this PR is slightly a mess with all the commits

File: imap_processing/cdf/global_attrs.py
Comment (maxinelasp) [reply imap_processing/cdf/global_attrs.py]: Yes

File: imap_processing/mag/l0/decom_mag.py
Comment (maxinelasp) [reply imap_processing/mag/l0/decom_mag.py:46]: I'd have to loop through twice. Once to get the list of decommed packets from decom_packets, then again to go through that list and apply the MagL0 code to it. One alternative might be to write an optional parameter in decom_packets that includes a dataclass to apply ... Since Python's typing system is so loose, you could then return either a list of packets or a list of, for example, MagL0 objects. Might be worth doing now that more people are using the dataclasses...?

File: imap_processing/mag/l0/decom_mag.py
Comment (maxinelasp) [reply imap_processing/mag/l0/decom_mag.py]: That is a good question that I don't fully remember the answer too. I think this is a variable coordinate that corresponds to the length of the vector (for each time stamp, there should be a vector of length 4 - x, y, z, and sensor range) but for these L0 RAW CDF files, there is instead a long string of bytes. So I think this was mostly me testing that I could get that 2d array in the CDF, but this format isn't correct for this output.

File: imap_processing/mag/l0/decom_mag.py
Comment (greglucas) [reply imap_processing/mag/l0/decom_mag.py]: I think you might need to add the new packets you had from one of the other commits, or update the XTCE definition to avoid the `ReadError`s.
https://github.com/IMAP-Science-Operations-Center/imap_processing/actions/runs/7920476930/job/21623597869#step:6:145

File: imap_processing/cdf/global_attrs.py
Comment (greglucas) [reply imap_processing/cdf/global_attrs.py]: Thanks for sending the examples to Bryan and following up with him. It seems like it is acceptable to do it this way and there is not necessarily a standard. We can update this at a later time if need be as well.

File: imap_processing/mag/l0/decom_mag.py
Comment (sdhoyt) [reply imap_processing/mag/l0/decom_mag.py:46]: You wouldn't know what APIDs are in the file to pass in a data class though, so you'd probably still have to loop through twice. For example, if the file contains science and housekeeping packets, you would need to pass in multiple data classes and have a way to match up the APIDs. I might be missing something here though. But I do see what you're saying about the way you're doing it now avoids the need to loop through the packets twice.

Comment (maxinelasp) [reply imap_processing/mag/l0/decom_mag.py:46]: Ah yeah... that's a good point. You could probably do it with some nasty dictionary shenanigans... but I hate nasty dictionary shenanigans! Personally, I think it's fine to just have every instrument do the opening and reading themselves, since there are special cases (dataclasses) and weird scenarios (what you said)

File: imap_processing/mag/l0/decom_mag.py
Comment (maxinelasp) [reply imap_processing/mag/l0/decom_mag.py]: I addressed this issue by updating the test files AND fixed the XTCE definition!

**Comment by maxinelasp** [comment]
Could I get another review @greglucas  or @sdhoyt with these new updates? Thanks for your prior comments!

File: imap_processing/cdf/global_attrs.py
Comment (bryan-harter) [reply imap_processing/cdf/global_attrs.py]: Eric Christian pointed out to me that there's apparently no such thing as an "Instrument PI" on IMAP, they are "Instrument Leads".  So PI name would probably just be reserved for Dave.  I'll get confirmation later if we need another global attribute for Instrument Leads

File: imap_processing/tests/mag/test_mag_decom.py
Comment (greglucas) [inline imap_processing/tests/mag/test_mag_decom.py]: I think you could leave this as a path being passed in so you don't have to cast to string in all of these.

File: imap_processing/tests/mag/test_mag_decom.py
Comment (greglucas) [inline imap_processing/tests/mag/test_mag_decom.py]:
```suggestion
    for index, test in enumerate(l0):
```
and remove the index += at the end of the loop

File: imap_processing/tests/mag/test_mag_decom.py
Comment (greglucas) [inline imap_processing/tests/mag/test_mag_decom.py]:
```suggestion
```

**Review by greglucas — APPROVED** [review]
Looks good to me! Some minor nits, but looks good overall. Thanks for updating the tests and removing the try/except, that is satisfying to see :)

File: imap_processing/cdf/global_attrs.py
Comment (bryan-harter) [reply imap_processing/cdf/global_attrs.py:8]: Eventually we'll probably want to add an attribute for the spase_DatasetResourceID, but it's not actually necessary.  We can discuss the SPASE requirement more at the SDWG tomorrow, but the internal format of our CDF files doesn't need to be anything special to comply to SPASE

File: imap_processing/cdf/global_attrs.py
Comment (maxinelasp) [reply imap_processing/cdf/global_attrs.py]: Sounds good. I'll remove this change from the PR and leave it as just Dave.

**Pull request merged**
